testdriver/testdriver.go

```diff
@@ -538,46 +525,148 @@ func (d *TestDriver) listenDriver() error {
 	}
+	return 0
+}
+
+func VimExprWait() (exitCode int) {
+	defer cleanUp(&exitCode)
 
 	fs := flag.NewFlagSet("vim", flag.PanicOnError)
-	bang := fs.Bool("bang", false, "expect command to fail")
-	indent := fs.Bool("indent", false, "pretty indent resulting JSON")
+	noindent := fs.Bool("noindent", false, "do not pretty indent resulting JSON")
 	stringout := fs.Bool("stringout", false, "print resulting string rather than JSON encoded version of string")
-
-	log("starting vim driver client and parsing flags...")
+	fWait := fs.String("wait", "", "retry (with exp backoff) until this time period has elapsed")
 
 	fs.Parse(os.Args[1:])
 	args := fs.Args()
+	if len(args) < 2 {
+		ef("invalid arguments")
+	}
+	fileName := args[0]
+	want, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		ef("failed to read %v: %v", fileName, err)
+	}
+	// Now repeatedly evaluate the provided expression
+	// with exp backoff until we match want. Abort on any errors
+	// because this might indicative of a greater problem.
+	vimCmd := append([]string{"expr"}, args[1:]...)
+	if *fWait == "" {
+		fWait = &DefaultErrLogMatchWait
+	}
+	wait, err := time.ParseDuration(*fWait)
+	if err != nil {
+		ef("failed to parse -wait duration %q: %v", *fWait, err)
+	}
+	strategy := retry.LimitTime(wait,
+		retry.Exponential{
+			Initial: 100 * time.Millisecond, // be slightly less aggressive with Vim
+			Factor:  1.5,
+		},
+	)
+	var got *bytes.Buffer
+	for a := retry.Start(strategy, nil); a.Next(); {
+		resp, err := vim(vimCmd)
+		if err != nil {
+			ef("%v", err)
+		}
+		if resp[0] != "" {
+			ef("protocol error: %v", resp[0])
+		}
+		vimResp := resp[1].([]interface{})
+		if vimResp[0] != "" {
+			ef("vim error: %v", vimResp[0])
+		}
+		if len(vimResp) != 2 {
+			ef("expected return value from Vim; saw len(vimResp) = %v", len(vimResp))
+		}
+		got = new(bytes.Buffer)
+		if *stringout {
+			switch v := vimResp[1].(type) {
+			case string:
+				got.WriteString(v)
+			default:
+				ef("response type is %T, not string", vimResp[1])
+			}
+		} else {
+			enc := json.NewEncoder(got)
+			if !*noindent {
+				enc.SetIndent("", "  ")
+			}
+			if err := enc.Encode(vimResp[1]); err != nil {
+				ef("failed to format output of JSON: %v", err)
+			}
+		}
+		if bytes.Equal(want, got.Bytes()) {
+			return 0
+		}
+	}
+	// We failed to find a match. Show diff and return an error
+	fmt.Fprint(os.Stderr, textutil.Diff(got.String(), string(want)))
+
+	return 1
+}
+
+// vim makes a call to Vim. args[0:2] are expected to be quoted
+// strings. args[2:] are expected to be string literals representing
+// JSON. resp is a tuple such that:
+//
+// resp[0] is a string representing any protocol error
+// resp[1] is a []interface such that:
+//   resp[1][0] is a string representing any Vim error
+//   resp[1][1] (if supplied) is the value returned by Vim
+func vim(args []string) (resp [2]interface{}, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
```

Comment:
What if r is not an error, as would be the case from ef?

Reply:
Thanks good catch. I missed changed `ef` to panic with `fmt.Errorf`